concordium-consensus/src/Concordium/GlobalState/Persistent/Accounts.hs

```diff
@@ -256,24 +415,70 @@ updateAccountsAtIndex' fupd ai = fmap snd . updateAccountsAtIndex fupd' ai
+--  There are no guarantees of the order of the list.
+allAccounts :: (SupportsPersistentAccount pv m) => Accounts pv -> m [(AccountAddress, AccountIndex)]
+allAccounts accounts = do
+    mDiffMap <- liftIO $ readIORef (accountDiffMapRef accounts)
+    case mDiffMap of
+        Absent -> LMDBAccountMap.getAllAccounts $ (AccountIndex . L.size) (accountTable accounts) - 1
+        Present accDiffMap -> do
+            -- Get all persisted accounts from the account map up to and including the last account of the account table minus what we found the in the difference map.
+            flattenedDiffMapAccounts <- DiffMap.flatten accDiffMap
+            persistedAccs <- LMDBAccountMap.getAllAccounts $ (AccountIndex . L.size) (accountTable accounts) - (1 + fromIntegral (length flattenedDiffMapAccounts))
+            return $! persistedAccs <> flattenedDiffMapAccounts
+
 -- | Get a list of all account addresses.
-accountAddresses :: (MonadBlobStore m) => Accounts pv -> m [AccountAddress]
-accountAddresses = AccountMap.addresses . accountMap
+--  There are no guarantees of the order of the list. This is because the resulting list is potentially
+--  a concatenation of two lists of account addresses.
+accountAddresses :: (SupportsPersistentAccount pv m) => Accounts pv -> m [AccountAddress]
+accountAddresses accounts = map fst <$> allAccounts accounts
 
 -- | Serialize accounts in V0 format.
 serializeAccounts :: (SupportsPersistentAccount pv m, MonadPut m) => GlobalContext -> Accounts pv -> m ()
-serializeAccounts cryptoParams accts = do
-    liftPut $ putWord64be $ L.size (accountTable accts)
-    L.mmap_ (serializeAccount cryptoParams) (accountTable accts)
+serializeAccounts cryptoParams Accounts{..} = do
+    liftPut $ putWord64be $ L.size accountTable
+    L.mmap_ (serializeAccount cryptoParams) accountTable
 
 -- | Fold over the account table in ascending order of account index.
 foldAccounts :: (SupportsPersistentAccount pv m) => (a -> PersistentAccount (AccountVersionFor pv) -> m a) -> a -> Accounts pv -> m a
 foldAccounts f a accts = L.mfold f a (accountTable accts)
 
--- | Fold over the account table in ascending order of account index.
+-- | Fold over the account table in descending order of account index.
 foldAccountsDesc :: (SupportsPersistentAccount pv m) => (a -> PersistentAccount (AccountVersionFor pv) -> m a) -> a -> Accounts pv -> m a
 foldAccountsDesc f a accts = L.mfoldDesc f a (accountTable accts)
 
+-- | Initialize the LMDB account map if it is not already.
+--  This puts in all accounts from the account table of the provided block state into the account map.
+--  If there already are accounts present in the account map, then we check that the size of the account map
+--  corresponds with number of accounts in the account table.
+--  If the number of accounts in the account table and account map matches, then this function does nothing.
+--  If they do not match, wipe the account map and recreate it from the account table.
```

Comment:
```suggestion
--  If the account map has fewer accounts than the provided account table, the account map is
--  wiped and repopulated from the account table. Otherwise, the account map is unchanged.
```